20231018_signal_cmb/maps.py

```diff
@@ -14,78 +15,133 @@
     lb = None
-def _build_maps(idx, beam_amin: float, nside:int, job='TQU'):
-    assert os.path.exists(path2cmb%idx), 'cmb alms not found at ' + path2cmb%idx
-    assert job.upper() in ['T', 'QU', 'TQU']
-    lmax = 4096
+
+def _slice_alms(teb, lmax_new):
+    """Returns the input teb alms sliced to the new lmax.
+
+    teb(numpy array): input teb alms
+    lmax_new(int): new lmax
+    """
+
+    lmax = hp.Alm.getlmax(len(teb[0]))
+    if lmax_new > lmax:
+        raise ValueError('lmax_new must be smaller or equal to lmax')
+    elif lmax_new == lmax:
+        return teb
+    else:
+        teb_new = np.zeros((3, hp.Alm.getsize(lmax_new)), dtype=teb.dtype)
+        indices_full = hp.Alm.getidx(lmax,*hp.Alm.getlm(lmax_new))
+        indices_new = hp.Alm.getidx(lmax_new,*hp.Alm.getlm(lmax_new))
+        teb_new[:,indices_new] = teb[:,indices_full]
+        return teb_new
+
+def _build_maps(idx, beam_amin: float, nside: int, job="TQU", lmax=4096, coord=None):
+    """Returns CMB-S4/LiteBIRD simulated T, Q and U maps.
+
+    idx(int): simulation index
+    beam_amin(float): beam width in arcmin
+    nside(int): healpix nside of the instrument model
+    job(str, optional): one of 'T', 'QU' or 'TQU' (default), for temperature-only, pol-only, or all three maps
+    lmax(int, optional): maximum multipole of the alm (default 4096)
+    coord(list, optional): list of strings, one of 'C' or 'G' (default), for coordinate system of the output maps
+
+    returns:
+        numpy array of shape (ncomp, npix), with ncomp set by 'job' and npix by the nside of the instrument model
+        CMB maps in uK
+
+    """
+    assert os.path.exists(path2cmb % idx), "cmb alms not found at " + path2cmb % idx
+    assert job.upper() in ["T", "QU", "TQU"]
+    if coord:
+        assert type(coord) == list, "coord should be a list of strings"
+        rot = hp.Rotator(coord=coord)
 
     # Builds pixel and beam:
-    bl_T, bl_E, bl_B, _ = hp.gauss_beam(beam_amin/ 180 / 60 * np.pi, pol=True, lmax=lmax).T
+    bl_T, bl_E, bl_B, _ = hp.gauss_beam(
+        beam_amin / 180 / 60 * np.pi, pol=True, lmax=lmax
+    ).T
 
     # Additional window function ?
     # pw_T, pw_P = hp.pixwin(nside, lmax=lmax, pol=True)
-    pw_T, pw_P = 1., 1.
+    pw_T, pw_P = 1.0, 1.0
 
     # load lencmbs and apply the transfer function
-    teb = np.load(path2cmb%idx)
+    teb = _slice_alms(np.load(path2cmb%idx),lmax)
     ncomp = ('T' in job.upper()) + 2 * ('QU' in job.upper())
     maps = np.empty((ncomp, hp.nside2npix(nside)), dtype=float)
     if 'T' in job.upper():
         hp.almxfl(teb[0], bl_T * pw_T, inplace=True)
+        if coord:
+            rot.rotate_alm(teb[0], inplace=True)
         maps[0] = hp.alm2map(teb[0], nside)
     if 'QU' in job.upper():
         hp.almxfl(teb[1], bl_E * pw_P, inplace=True)
         hp.almxfl(teb[2], bl_B * pw_P, inplace=True)
+        if coord:
+            rot.rotate_alm(teb[1], inplace=True)
```

Comment:
I think you need to rotate TEB all together, so polarization is handled properly by HEALPix